dds/DCPS/XTypes/DynamicType.cpp

```diff
@@ -0,0 +1,138 @@
+#include "DCPS/DdsDcps_pch.h"
+#include "DynamicType.h"
+#include "MemberDescriptor.h"
+#include "TypeDescriptor.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+DynamicType::DynamicType()
+  : descriptor_(DCPS::make_rch<XTypes::TypeDescriptor>())
+{}
+
+void DynamicType::get_descriptor(TypeDescriptor_rch& descriptor) const
+{
+  descriptor = descriptor_;
+}
+
+TypeDescriptor_rch DynamicType::get_descriptor()
+{
+  return descriptor_;
+}
+
+DCPS::String DynamicType::get_name() const
+{
+  return descriptor_->name;
+}
+
+TypeKind DynamicType::get_kind() const
+{
+  return descriptor_->kind;
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_name(DynamicTypeMember_rch& member, const DCPS::String& name) const
+{
+  const DynamicTypeMembersByName::const_iterator pos = member_by_name.find(name);
+  if (pos != member_by_name.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members_by_name(DynamicTypeMembersByName& member)
+{
+  member = member_by_name;
+}
+
+DDS::ReturnCode_t DynamicType::get_member(DynamicTypeMember_rch& member, MemberId id) const
+{
+  DynamicTypeMembersById::const_iterator pos = member_by_id.find(id);
+  if (pos != member_by_id.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members(DynamicTypeMembersById& member)
+{
+  member = member_by_id;
+}
+
+ACE_CDR::ULong DynamicType::get_member_count() const
+{
+  return CORBA::ULong(member_by_index.size());
```

Comment:
Use a consistent typedef for this type.  `ACE_CDR::ULong` has the advantage of not being tied to any specific IDL mapping spec.